pilot/pkg/grpc/tls.go

```diff
@@ -0,0 +1,111 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpc
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net"
+	"os"
+	"strings"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"istio.io/istio/security/pkg/pki/util"
+	"istio.io/pkg/log"
+)
+
+// TLSOptions include TLS options that a grpc client uses to connect with server.
+type TLSOptions struct {
+	RootCert      string
+	Key           string
+	Cert          string
+	ServerAddress string
+	SAN           string
+}
+
+func getTLSDialOption(opts *TLSOptions) (grpc.DialOption, error) {
```

Comment:
```suggestion
func getTLSDialOption(opts TLSOptions) (grpc.DialOption, error) {
```

No need for a pointer?

Reply:
I meant to. For got to do

Reply:
This is because we need to distinguish using tls or plaintext from the caller.